Feature: disallow version overwrites by policy

emauton opened the pull request:
Prior to this change, re-uploading a package with the same version
results in the old version being overwritten.

In some environments (e.g. with packages being automatically uploaded
from "green" builds of `master`), this can result in programmer error -
forgetting to change `setup.py` versions - changing the "version of
record" in the local `pyshop`.

This change add a new administrative policy option

    pyshop.upload.never_overwrite

which results in re-uploads raising a HTTP Conflict Error.

This change also introduces `pyfakefs` to test cases, which means that
they no longer rely upon real `/tmp` or have filesystem dependencies
across tests.

emauton committed:
Feature: disallow version overwrites by policy

Prior to this change, re-uploading a package with the same version
results in the old version being overwritten.

In some environments (e.g. with packages being automatically uploaded
from "green" builds of `master`), this can result in programmer error -
forgetting to change `setup.py` versions - changing the "version of
record" in the local `pyshop`.

This change add a new administrative policy option

    pyshop.upload.never_overwrite

which results in re-uploads raising a HTTP Conflict Error.

This change also introduces `pyfakefs` to test cases, which means that
they no longer rely upon real `/tmp` or have filesystem dependencies
across tests.

emauton commented:
Just finishing up for the day. Any thoughts on the PR appreciated, and I'll take a look at the Travis errors in the morning. Thanks!

mardiros commented:
It appears you break the tests suite ?

emauton commented:
Yes, I'm a little unclear on why `pyfakefs` is breaking under Travis but not under local virtualenvs, but I'll take a look today.

emauton committed:
Add deps for non-PyPI pyfakefs

Travis runs using `nosetests`. While `python setup.py test` works fine,
something about interaction with `nose` is causing trouble for the 3.2
version of `pyfakefs` available on PyPI.

It seems 3.3 from `pyfakefs` Github works fine; opened an issue asking
about it at https://github.com/jmcgeheeiv/pyfakefs/issues/304

emauton commented:
I've cleared up the build problem with `pyfakefs`, but arguably in a less than ideal way. :o)

See notes at https://github.com/mardiros/pyshop/pull/86/commits/675f9243f91d817e1d27d6f2f26bf20fb1f42157

and let me know what you think. Thanks!

mardiros commented:
I like the idea of the setting to disallow override.

I didn't know about pyfakefs. I think we can wait a bit to know if its author release a new version of it
to avoid a git+http dependency.

The dependency has to be specified here :

https://github.com/mardiros/pyshop/blob/master/setup.py#L36

and must be a packaged version.

emauton commented:
Yeah, funnily enough I'm working with `pyshop` precisely to stop using private git dependencies in my company's setup. :o)

I'll check back after we see what happens on https://github.com/jmcgeheeiv/pyfakefs/issues/304.

Thanks!

emauton committed:
Remove `git+https` deps for `pyfakefs`

No longer necessary as the latest `pyfakefs` has been pushed to PyPI per
  https://github.com/jmcgeheeiv/pyfakefs/issues/304

emauton commented:
Looks like we're good to go here with latest `pyfakefs`.

emauton commented:
Neat, thanks for the merge & for `pyshop`. :o)
